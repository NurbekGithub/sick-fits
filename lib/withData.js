import withApollo from 'next-with-apollo';
import ApolloClient from 'apollo-boost';
import { endpoint, prodEndpoint } from '../config';
import { LOCAL_STATE_QUERY } from '../components/Cart';

export function dump(data) {
  return JSON.stringify(data, null, 2);
}

function createClient({ headers }) {
  return new ApolloClient({
    uri: process.env.NODE_ENV === 'development' ? endpoint : prodEndpoint,
    request: operation => {
      operation.setContext({
        fetchOptions: {
          credentials: 'include',
        },
        headers,
      });
    },
    clientState: {
      resolvers: {
        Mutation: {
          toggleCart(_, variables, { cache }) {
            // 1. read the cartOpen value from the cache
            const { cartOpen } = cache.readQuery({ query: LOCAL_STATE_QUERY });
            // 2. write the cartOpen value to the cache
            const data = {
              data: { cartOpen: !cartOpen }
            };
            cache.writeData(data);
            return data;
          }
        }
      },
      defaults: {
        cartOpen: false
      }
    }
  });
}

export default withApollo(createClient);
