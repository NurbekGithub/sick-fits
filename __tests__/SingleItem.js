import { MockedProvider } from 'react-apollo/test-utils';
import waait from 'waait';
import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import {fakeItem} from '../lib/testUtils';
import SingleItem, { SINGLE_ITEM_QUERY } from '../components/SingleItem';

describe('<SingleItem />', () => {
  const mocks = [
    {
      // when someone makes the request
      request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
      // they get this response
      result: {
        data: {
          item: fakeItem()
        }
      }
    }
  ]
  it('should render and show loading', () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id='123'/>
      </MockedProvider>
    );
    expect(wrapper.text()).toBe('Loading...');
  });
  
  it('should match snap', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id='123'/>
      </MockedProvider>
    );
    await waait();
    wrapper.update(); 
    expect(toJSON(wrapper.find('h2'))).toMatchSnapshot();  
    expect(toJSON(wrapper.find('img'))).toMatchSnapshot();  
    expect(toJSON(wrapper.find('p'))).toMatchSnapshot();  
  });

  it('should error with no item found', async () => {
    const mocks = [
      {
        request: { query: SINGLE_ITEM_QUERY, variables: { id: '123' } },
        result: { errors: [{message: 'Items Not Found'}] }
      }
    ]
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <SingleItem id='123'/>
      </MockedProvider>
    );
    await waait();
    wrapper.update();
    const item = wrapper.find('[data-test="graphql-error"]');
    expect(item.text()).toContain('Items Not Found');
    expect(toJSON(item)).toMatchSnapshot();
  })
})