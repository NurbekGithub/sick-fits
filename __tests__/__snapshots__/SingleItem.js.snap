// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItem /> should error with no item found 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Items Not Found
</p>
`;

exports[`<SingleItem /> should match snap 1`] = `
<h2>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleItem /> should match snap 2`] = `
<img
  alt="dogs are best"
  src="dog.jpg"
/>
`;

exports[`<SingleItem /> should match snap 3`] = `
<p>
  dogs
</p>
`;
